mixer/adapter/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,134 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate $GOPATH/src/istio.io/istio/bin/mixer_codegen.sh -f mixer/adapter/cloudwatch/config/config.proto
+
+package cloudwatch
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/service/cloudwatch/cloudwatchiface"
+
+	"istio.io/istio/mixer/adapter/cloudwatch/config"
+
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+)
+
+type (
+	builder struct {
+		adpCfg      *config.Params
+		metricTypes map[string]*metric.Type
+	}
+
+	// Handler holds data for the cloudwatch adapter handler
+	Handler struct {
+		metricTypes map[string]*metric.Type
+		env         adapter.Env
+		cfg         *config.Params
+		cloudwatch  cloudwatchiface.CloudWatchAPI
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &Handler{}
+
+///////////////// Configuration-time Methods ///////////////
+
+// NewHandler initializes a cloudwatch handler
+func NewHandler(metricTypes map[string]*metric.Type, env adapter.Env, cfg *config.Params, cloudwatch cloudwatchiface.CloudWatchAPI) adapter.Handler {
+	return &Handler{metricTypes: metricTypes, env: env, cfg: cfg, cloudwatch: cloudwatch}
+}
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	cloudwatch := NewCloudWatchClient()
+	return NewHandler(b.metricTypes, env, b.adpCfg, cloudwatch), nil
+}
+
+// adapter.HandlerBuilder#SetAdapterConfig
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adpCfg = cfg.(*config.Params)
+}
+
+// adapter.HandlerBuilder#Validate
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	cfg := b.adpCfg
+
+	// validate that instance and handler contain the same metrics
+	for _, v := range cfg.GetMetricInfo() {
+		if v.GetStorageResolution() != 1 && v.GetStorageResolution() != 60 {
+			ce = ce.Appendf("storageResolution", "Storage resolution should be either 1 or 60")
+		}
+
+		//TODO: validate cw unit
+		if len(v.GetUnit()) == 0 {
+			ce = ce.Appendf("unit", "Cloudwatch metric unit must not be null")
+		}
+	}
+
+	return
+}
+
+// metric.HandlerBuilder#SetMetricTypes
+func (b *builder) SetMetricTypes(types map[string]*metric.Type) {
+	b.metricTypes = types
+}
+
+// HandleMetric validates received instances and sends them to cloudwatch
+func (h *Handler) HandleMetric(ctx context.Context, insts []*metric.Instance) error {
+	validInsts := getValidMetrics(h, insts)
+	return handleValidMetrics(h, validInsts)
+}
+
+func getValidMetrics(h *Handler, insts []*metric.Instance) []*metric.Instance {
+	validInsts := make([]*metric.Instance, 0, len(insts))
+
+	for _, inst := range insts {
+		if _, ok := h.metricTypes[inst.Name]; !ok {
+			_ = h.env.Logger().Errorf("There is no type found for instance %s", inst.Name)
+			continue
+		}
+
+		// check that every instance metric has a corresponding cloudwatch handler metric
```

Comment:
This check should be done in the Validate method instead, it doesn't need to happen on every request.

The idea with the design is that all validation is performed at config time, such that at runtime there is little overhead at possible as we process requests.

Reply:
Makes sense. Is there any way to get instances in validate?

Reply:
No instances during Validate.

Validate executes at config time, just to make sure something the user wrote is valid. There is no traffic flowing at that point, so there's nothing to create instances from. You're given type information that describes the full set of possible instance types you'll get at runtime. So you can validate to make sure you'll be happy with the kind of data Mixer will hand your adapter at runtime.

So if you have any limits, like max dimensions, you want those to be reflected in Validate.

In the common case, the only reason why a HandleXXX call should ever return an error is because network connectivity to the backend failed. Other error causes should be found during Validate.

Reply:
Gotcha. So I guess there is no way to validate that the configuration in handler config and instance config matches then?

Reply:
Currently I'm validating that instance names defined in instance config have matching config in handler config.

Reply:
This is Mixer invariant, we guarantee consistency.

- Mixer will _always_ call Validate to ensure config is valid before calling Build.

- Mixer will _always_ give you all the type descriptors up front before calling Validate

- Mixer will _never_ call your HandleXXX function with instances that reference types which weren't given to you before Build.

In other words, adapter code can assume everything is sane around them. So they check their config in Validate and everything else is taken care of for them.